app/src/main/java/com/samco/trackandgraph/antlr/evaluation/Evaluation.kt

```diff
@@ -0,0 +1,99 @@
+package com.samco.trackandgraph.antlr.evaluation
+
+
+import com.samco.trackandgraph.antlr.ast.*
+import com.samco.trackandgraph.antlr.parsing.DatatransformationFunctionParserFacade
+import com.samco.trackandgraph.database.entity.DataPoint
+
+fun inferDatatype(data: List<DataPoint>) : DataType {
+    return DataType.NUMERICAL
+}
+
+class EvaluationModel {
+
+    fun run(code: String, inputFeatures: Map<String, List<DataPoint>> = emptyMap()) : Map<String, Value> {
+        val parseResult = DatatransformationFunctionParserFacade.parse(code, inputFeatures.keys)
+        when (parseResult.errors.size) {
+            0 -> {}
+            1 -> throw parseResult.errors[0]
+            else -> throw ListOfErrors(parseResult.errors)
+        }
+
+        return this.evaluate(parseResult.root!!, inputFeatures)
+    }
+
+    private fun evaluate(
+        parsedTree: DatatransformationFunction,
+        inputFeatures: Map<String, List<DataPoint>> = emptyMap<String, List<DataPoint>>()
+    ) : Map<String, Value> {
+        var context : Map<String, Value> = inputFeatures.mapValues { list -> DatapointsValue(list.value, inferDatatype(list.value)) }
+        for (statement in parsedTree.statements) {
+            context = when (statement) {
+                is VarDeclaration -> statement.evaluate(context)
+                is Assignment -> statement.evaluate(context)
+                is Print -> statement.evaluate(context)
+                else -> throw UnsupportedOperationException(this.javaClass.canonicalName)
+            }
+
+        }
+
+        return context
+    }
+}
+
+fun Expression.evaluate(context: Map<String, Value>) : Value {
+    try {
+        return when (this) {
+            is IntLit -> NumberValue(this.value.toInt())
+            is DecLit -> NumberValue(this.value.toDouble())
+            is StringLit -> StringValue(this.value)
+            is TimeperiodLit -> TimeValue(this.context)
+
+            is MultiplicationExpression -> this.left.evaluate(context) * this.right.evaluate(context)
+            is DivisionExpression       -> this.left.evaluate(context) / this.right.evaluate(context)
+            is SumExpression            -> this.left.evaluate(context) + this.right.evaluate(context)
+            is SubtractionExpression    -> this.left.evaluate(context) - this.right.evaluate(context)
+
```

Comment:
For expressions we have to recursively go down and evaluate their parts. Note that this concise notation is the result of the whole Value override abstraction thing